### PR TITLE
New pfRemainingCharsCount directive

### PR DESCRIPTION
**Pull request by dtaylor113**
New pfRemainingCharsCount directive counts remaining characters for textareas and inputs.  Ability to specify when to warn user when certain number of characters remain.  Also to block input after maximum number of characters reached.

image

### CODE REVIEW

File: src/form/remaining-chars-count/remaining-chars-count.directive.js [hunk 1]
```diff
@@ -0,0 +1,177 @@
+/**
+ * @ngdoc directive
+ * @name patternfly.form.directive:pfRemainingCharsCount
+ *
+ * @description
+ *   Directive for showing a characters remaining count and triggering warning and error
+ *   behavior when passing specified thresholds.
+ *
+ * @param {string} ng-model The scope model variable which contains the intial text for the text field.  Required, but</br>
```
Comment (jeff-phillips-18) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: Typo on 'intial text'

File: src/form/remaining-chars-count/remaining-chars-count.directive.js [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+ * can be an emptly string ("").
+ * @param {string} count-fld The id of the field to display the 'characters-remaining' count.
```
Comment (jeff-phillips-18) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: Better to not abbreviate here, count-fld -> count-field

File: src/form/remaining-chars-count/remaining-chars-count.directive.js [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+ * @param {string} chars-max-limit Number representing the maximum number of characters to allow before dispatching a<br/>
```
Comment (jeff-phillips-18) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: how about  'max-chars' instead?

File: src/form/remaining-chars-count/remaining-chars-count.directive.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+ * 'overCharsMaxLimit' event.   When the number of characters falls below <code>chars-max-limit</code>, a 'underCharsMaxLimit'<br/>
+ * event is dispatched.
+ * @param {string} chars-warn-remaining Number of remaining characters to warn upon.  The 'chars-warn-remaining-pf'<br/>
+ * class will be applied to the <code>count-fld</code> when the remaining characters is less than the<br/>
+ * <code>chars-warn-remaining</code> threshold.  When/if the number of remaining characters becomes greater than the<br/>
+ * <code>chars-warn-remaining</code> threshold, the 'chars-warn-remaining-pf' class is removed from the <code>count-fld</code> field.
```
Comment (jeff-phillips-18) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: This information would be better as part of the overall directive description.

File: src/form/remaining-chars-count/remaining-chars-count.directive.js [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+     <style>
+       textarea {
+         resize: none;
+       }
+     </style>
+
+     <div class="container">
+       <strong>Max Limit: 20, Warn When 5 Or less Remaining, Disable Button After Max Limit</strong>
+       <div class="row">
```
Comment (jeff-phillips-18) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: Nit - don't use camel case here.

File: src/form/remaining-chars-count/remaining-chars-count.directive.js [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+           $scope.$digest();
+         }
+     });
+
+     // 'tfId' will equal the id of the text area/input field which
+     // triggered the event
+     $scope.$on('underCharsMaxLimit', function (event, tfId) {
+         if($scope.charsMaxLimitExceeded){
+           $scope.charsMaxLimitExceeded = false;
+           $scope.$digest();
```
Comment (jeff-phillips-18) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: Why is this necessary?

File: src/form/remaining-chars-count/remaining-chars-count.directive.js [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    require: 'ngModel',
+    scope: {
+      ngModel: "="
+    },
+    link: function ($scope, $element, $attributes) {
+      var charsMaxLimit = $attributes.charsMaxLimit;
+      var charsWarnRemaining = $attributes.charsWarnRemaining;
+      var countRemainingFld = angular.element(document.getElementById($attributes.countFld));
+      var blockInputAtMaxLimit = ($attributes.blockInputAtMaxLimit === 'true');
+      var charsRemainingFn = function () {
```
Comment (jeff-phillips-18) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: how about checkCharactersRemaining?

File: src/form/remaining-chars-count/remaining-chars-count.directive.js [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        charsRemainingFn();
+      });
+
+      $element.bind('keypress', function (event) {
+        // Once the charsMaxLimit has been met or exceeded, prevent all keypresses from working
+        if (blockInputAtMaxLimit && $element.val().length >= charsMaxLimit) {
+          // Except backspace
+          if (event.keyCode !== 8) {
+            event.preventDefault();
+          }
```
Comment (jeff-phillips-18) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: Does this also prevent delete? Does this handle cut/paste?

File: src/form/remaining-chars-count/remaining-chars-count.directive.js (hunk 3 above)
Comment (dtaylor113) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: These parameters match what we put into the base PF ref. impl.
Isn't 'chars-max-limit' more descriptive?

File: src/form/remaining-chars-count/remaining-chars-count.directive.js (hunk 5 above)
Comment (dtaylor113) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: Sorry, where are you saying?  For ng-controller?

File: src/form/remaining-chars-count/remaining-chars-count.directive.js (hunk 8 above)
Comment (dtaylor113) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: This does prevent the Delete key, but backspace delete is available.
This does handle cut & paste, both from keys and right-click.

File: src/form/remaining-chars-count/remaining-chars-count.directive.js (hunk 3 above)
Comment (jeff-phillips-18) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: just seems a bit superfluous

File: src/form/remaining-chars-count/remaining-chars-count.directive.js (hunk 5 above)
Comment (jeff-phillips-18) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: No, for the title. Just a nit. "Max limit: 20. Warning when 5 or less remaining. Disable button after max limit reached."

File: src/form/remaining-chars-count/remaining-chars-count.directive.js (hunk 2 above)
Comment (dtaylor113) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: These parameters match what we put into the base PF ref. impl.

File: src/form/remaining-chars-count/remaining-chars-count.directive.js (hunk 6 above)
Comment (dtaylor113) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: In those situations where there might be multiple textareas, it might be good to know which one dispatched the event.  In case they are listening on the body.

File: src/form/remaining-chars-count/remaining-chars-count.directive.js (hunk 5 above)
Comment (dtaylor113) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: Fixed

File: src/form/remaining-chars-count/remaining-chars-count.directive.js (hunk 7 above)
Comment (dtaylor113) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: Fixed

File: src/form/remaining-chars-count/remaining-chars-count.directive.js (hunk 4 above)
Comment (dtaylor113) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: Fixed

File: src/form/remaining-chars-count/remaining-chars-count.directive.js (hunk 2 above)
Comment (jeff-phillips-18) [inline src/form/remaining-chars-count/remaining-chars-count.directive.js]: OK

**Comment by jeff-phillips-18** [comment]
LGTM
